components/core/src/clp/streaming_archive/single_file_archive/writer.cpp

```diff
@@ -0,0 +1,268 @@
+        std::filesystem::path const& multi_file_archive_path,
+        segment_id_t next_segment_id
+) -> void;
+
+auto get_file_size_and_update_offset(std::filesystem::path const& file_path, uint64_t& offset)
+        -> void {
+    try {
+        auto size = std::filesystem::file_size(file_path);
+        offset += size;
+    } catch (std::filesystem::filesystem_error const& e) {
+        throw OperationFailed(
+                ErrorCode_Failure,
+                __FILENAME__,
+                __LINE__,
+                fmt::format("Failed to get file size: {}", e.what())
+        );
+    }
+}
+
+auto
+get_file_infos(std::filesystem::path const& multi_file_archive_path, segment_id_t next_segment_id)
+        -> std::vector<FileInfo> {
+    std::vector<FileInfo> files;
+    uint64_t offset = 0;
+
+    for (auto const& static_archive_file_name : cStaticArchiveFileNames) {
+        files.emplace_back(FileInfo{std::string(static_archive_file_name), offset});
+        get_file_size_and_update_offset(multi_file_archive_path / static_archive_file_name, offset);
+    }
+
+    std::filesystem::path segment_dir_path = multi_file_archive_path / cSegmentsDirname;
+
+    for (size_t i = 0; i < next_segment_id; ++i) {
+        auto const segment_id = std::to_string(i);
+        files.emplace_back(FileInfo{segment_id, offset});
+        get_file_size_and_update_offset(segment_dir_path / segment_id, offset);
+    }
+
+    // Add sentinel indicating total size of all files.
+    files.emplace_back(FileInfo{"", offset});
+
+    // Decompression of large single-file archives will consume excessive memory since
+    // single-file archives are not split.
+    if (offset > cFileSizeWarningThreshold) {
+        SPDLOG_WARN(
+                "Single file archive size exceeded {}. "
+                "The single-file archive format is not intended for large archives, "
+                " consider using multi-file archive format instead.",
+                cFileSizeWarningThreshold
+        );
+    }
+
+    return files;
+}
+
+auto pack_single_file_archive_metadata(
+        ArchiveMetadata const& multi_file_archive_metadata,
+        std::filesystem::path const& multi_file_archive_path,
+        segment_id_t next_segment_id
+) -> std::stringstream {
+    SingleFileArchiveMetadata single_file_archive{
+            .archive_files = get_file_infos(multi_file_archive_path, next_segment_id),
+            .archive_metadata = multi_file_archive_metadata,
+            .num_segments = next_segment_id,
+    };
+
+    std::stringstream buf;
+    msgpack::pack(buf, single_file_archive);
+
+    return buf;
+}
+
+auto write_archive_header(FileWriter& archive_writer, size_t packed_metadata_size) -> void {
+    SingleFileArchiveHeader header{
+            .magic{},
+            .version = cArchiveVersion,
+            .metadata_size = packed_metadata_size,
+            .unused{}
+    };
+
+    static_assert(cUnstructuredSfaMagicNumber.size() == header.magic.size());
```

Comment:
Is there any chance we can get the size of magic without actually creating the header object?
imo, a better order would be:
1. static_assert the size equals expected.
2. Initialize header, and hopefully we don't need memcpy at end.